rfc/system/RFC-5659-execution-concurrency.md

```diff
@@ -0,0 +1,317 @@
+Therefore, this proposal only applies concurrency at the versioned launch plan level.
+
+If we wanted to support concurrency across launch plan versions:
+
+We could update usage like so
+
+```python
+my_lp = LaunchPlan.get_or_create(
+    name="my_serial_lp",
+    workflow=my_wf,
+    ...
+    concurrency=Concurrency(
+        max=1,  # defines how many executions with this launch plan can run in parallel
+        policy=ConcurrencyPolicy.WAIT  # defines the policy to apply when the max concurrency is reached
+        precision=ConcurrencyPrecision.LAUNCH_PLAN
+    )
+)
+```
+
+and by default, when the precision is omitted the SDK could register the launch plan using `ConcurrencyPrecision.LAUNCH_PLAN_VERSION`
+
+We could update the concurrency protobuf definition like so:
+```protobuf
+message Concurrency {
+   // Defines how many executions with this launch plan can run in parallel
+   uint32 max = 1;
+
+   // Defines how to handle the execution when the max concurrency is reached.
+   ConcurrencyPolicy policy = 2;
+
+   ConcurrencyLevel level = 3;
+}
+
+enum ConcurrencyLevel {
+  UNSPECIFIED = 0;
+  
+  // Applies concurrency limits across all launch plan versions.
+  LAUNCH_PLAN = 1;
+  
+  // Applies concurrency at the versioned launch plan level
+  LAUNCH_PLAN_VERSION = 2;
+}
+```
+
+
+We could add another index to the Executions table to include the launch plan named entity, that is the entry in the [NamedEntity table](https://github.com/flyteorg/flyte/blob/38883c721dac2875bdd2333f4cd56e757e81ea5f/flyteadmin/pkg/repositories/models/named_entity.go#L39-L42) corresponding to the launch plan project, domain & name
+
+In models/execution.go:
+```go
+
+type Execution struct {
+	
+	...
+	// Already exists
+	LaunchPlanID uint   `gorm:"index"`
+	// New field to make querying on the named entity 
+	LaunchPlanNamedEntityID uint `gorm:"index"`
+	// New field to make querying on concurrency policy by the reconciliation loop easier
+	ConcurrencyLevel uint32
+	
+}
+
+
+```
+
+Then the reconciliation loop would query executions in a non-terminal phase matching the launch plan named entity ID instead of LaunchPlanID based on the ConcurrencyLevel.
+
+```sql
+SELECT e.*
+FROM   executions AS e
+WHERE  ( launch_plan_named_entity_id, created_at ) IN (SELECT launch_plan_named_entity_id,
+                                                 Min(created_at)
+                                          FROM   executions
+                                          WHERE  phase = 'PENDING' AND concurrency_level = 2;
+                                          GROUP  BY launch_plan_named_entity_id); 
+```
+
+Note, in this proposal, registering a new version of the launch plan and setting it to active will determine the concurrency policy across all launch plan versions.
+
+#### Prior Art
+The flyteadmin native scheduler (https://github.com/flyteorg/flyte/tree/master/flyteadmin/scheduler) already implements a reconciliation loop to catch up on any missed schedules.
```

Comment:
exactly look at the prior art. The number of db queries is minimal. Actually one per cycle.

I would keep all running executions with concurrency_level set in memory and all lps with concurrency_level set in memory (only the concurrency policies)
We should periodically update these and its ok to be eventually consistent